readme.txt
=== Postie ===
Contributors: WayneAllen
Donate link: https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=HPK99BJ88V4C2
Author URI: http://allens-home.com/
Plugin URI: http://PostiePlugin.com/
Tags: e-mail, email
Requires at least: 3.0
Tested up to: 3.5.1
Stable tag: 1.5.7
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html

Postie allows you to blog via e-mail, including many advanced features not found in WordPress's default post by e-mail feature.

== Description ==
Postie offers many advanced features for posting to your blog via e-mail,
including the ability to assign categories by name, included pictures and
videos, and automatically strip off signatures. It also has support for both
imap and pop3, with the option for ssl with both.  For usage notes, see the
[other notes](other_notes) page

== Screenshots ==

1. Postie options (showing video and audio templates)
== Installation ==
* Either:
    * Put the postie.zip file in wp-content/plugins/ and unzip it
* Or:
    * Use the automatic installer (WP 2.7+)
* Login to WordPress as an administrator
* Goto the Plugins tab in the WordPress Admin Site
* Activate "Postie"
* Goto to the "Settings" tab and click on the sub-tab "Postie" to configure it.
* Make sure you enter the mailserver information correctly, including the type
  of connection and the port number. Common port configurations:
  * pop3: 110 
  * pop3-ssl: 995
  * imap: 143
  * imap-ssl: 993
* (Postie ignores the settings under Settings->Writing->Writing-by-Email)

= Automating checking e-mail =
WordPress cron (which Postie relies on) doesn't run unless a page is accessed on 
the site. So if you send an email, but nobody accesses the site for 3 days Postie 
won't be given the chance to fetch the email and publish the post.

To ensure that Postie runs smoothly on a low or no volume site you need to ensure 
that a page gets hit (any page is fine). Use something like cron + curl on Linux 
or install curl on Windows and use the Scheduled Tasks control panel. If you are 
using a hosting company that doesn't allow you access to cron you can use a service 
like SetCronJob.

By default, postie checks for new e-mail every 30 minutes. You can select from
a number of different checking intervals in the settings page, under the
Mailserver tab.

= Forcing e-mail check =
If you would prefer to have more fine-grained control of how postie checks
for mail, you can also set up a crontab. This is for advanced users only.

Setup a cronjob to pull down the get\_mail.php - note that every time you access
get\_mail.php Postie will run - it is like clicking Run Postie on the Admin screen.

= Linux =
If your site runs on a UNIX/linux server, and you have shell access, you can
enable mail checking using cron. 

Examples:

*/5 * * * * /usr/bin/lynx --source http://blog.robfelty.com/wp-content/plugins/postie/get\_mail.php >/dev/null 2>&1

This fetches the mail every five minutes with lynx 

*/10 * * * * /usr/bin/wget -O /dev/null http://blog.robfelty.com/wp-content/plugins/postie/get\_mail.php >/dev/null 2>&1

This fetches the mail every ten minutes with wget 

= Windows =
You will need to install [wget](http://gnuwin32.sourceforge.net/packages/wget.htm) or [curl](http://curl.haxx.se/dlwiz/)
Then use the Task Scheduler control panel to call wget or cron.

== Usage ==
= Specifying Beginning and Ending of Post =
* If you put in :start - the message processing won't start until it sees that string.
* If you put in :end - the message processing will stop once it sees that string.

= Post Status =
* Posts can have their status set to draft, publish, pending or private. This will override the Default Post Status set in the settings screen.
  *    status: private
  *    status: draft

= Post Date =
* Posts can have a specific publication date such as Apr 14, 2013. Relative dates like "tomorrow", "monday", "first day of next month" are also supported.
  *    date: date
  *    date: date time
* Posts can be delayed by adding a line with delayXdXhXm where X is a number.
  *    delay:1d - 1 day
  *    delay:1h - 1 hour
  *    delay:1m - 1 minute
  *    delay:1d2h4m - 1 day 2 hours 4m

= Comment Control =
* By putting comments:X in your message you can control if comments are allowed
   *   comments:0 - means closed
   *   comments:1 - means open
   *   comments:2 - means registered only
* Replying to an e-mail gets posted as a comment. 
  * For example, you e-mailed a post with the subject line "foo".
    If you then send an e-mail with the subject line "Re: foo", it will
    get posted as a comment to the "foo" post. This works by the subject
    line, so if you have two posts with titles "foo", then the comment
    will get placed in the more recent post.

= Post Excerpt =
* Custom excerpt
  * You can include a custom excerpt of an e-mail by putting it between
    :excerptstart and :excerptend
    * You can include images in the excerpt by using the shortcode #eimg1#,
      #eimg2# etc.

= Post type/format =
  You can specify the post type or format by including it as the first part of the subject.
  E.g. aside//real subject

= Categories =
* If you put a category name in the subject with a : it will be used
  as the category for the post
* If you put a category id number in the subject with a : it will
  be used as the category for the post
* If you put the first part of a category name it will be posted in
  the first category that the system finds that matches - so if you put

  Gen: New News

  The system will post that in General. Note you must turn on the "Match short category"
  setting for this to work (on by default).

* All of the above also applies if you put the category in brackets []
* Using [] or you can post to multiple categories at once

  Subject: [1] [Mo] [Br] My Subject

  On my blog it would post to General (Id 1), Moblog, and Brewing all at one time

* Using - or you can post to multiple categories at once

  Subject: -1- -Mo- -Br- My Subject

  On my blog it would post to General (Id 1), Moblog, and Brewing all at one time

= Tags =
* You can add tags by adding a line in the body of the message like so:
  tags: foo, bar
* You can also set a default tag to be applied if no tags are included.

= Image Handling =
Note you can only use this feature if your "Preferred Text Type" is set to "plain"

* Allows you to attach images to your email and automatically post
  them to your blog
* You can publish images in the text of your message by using #img1#
  #img2# - each one will be replaced with the HTML for the image
  you attached
* Captions - you can also add a caption like so:

    * #img1 caption='foo'#
    * #img2 caption='bar'#
  
  Or, if you use IPTC captions, this caption will be used  (adding a caption
  in many photo editing programs (for example Picasa), will add an IPTC caption)

  Note that the images are processed in the order they are attached as of version
  1.4.6.

* Image templates
  Postie now uses the default wordpress image template, but you can specify a
  different one if you wish.

  You can also specify a custom image template. I use the following custom
  template:

  `<div class='imageframe alignleft'><a href='{IMAGE}'><img src="{THUMBNAIL}"
  alt="{CAPTION}" title="{CAPTION}" 
  class="attachment" /></a><div
  class='imagecaption'>{CAPTION}</div></div>`
     
    * {CAPTION} gets replaced with the caption you specified (if any)
    * {FILELINK} gets replaced with the url to the media
    * {FILENAME} gets replaced with the name of the attachment from the email
    * {FULL} same as {FILELINK}
    * {HEIGHT} gets replaced with the height of the photo
    * {ID} gets replaced with the post id
    * {IMAGE} same as {FILELINK}
    * {LARGEHEIGHT} gets replaced with the height of a large image
    * {LARGEWIDTH} gets replaced with the width of a large image
    * {LARGE} gets replaced with the url to the large-sized image
    * {MEDIUMHEIGHT} gets replaced with the height of a medium image
    * {MEDIUMWIDTH} gets replaced with the width of a medium image
    * {MEDIUM} gets replaced with the url to the medium-sized image
    * {PAGELINK} gets replaced with the URL of the file in WordPress
    * {POSTTITLE} gets replaced with the post title (subject)
    * {RELFILENAME} gets replaced with the relative path to the full-size image
    * {THUMBHEIGHT} gets replaced with the height of a thumbnail image
    * {THUMB} gets replaced with the url to the thumbnail image
    * {THUMBNAIL} same as {THUMB}
    * {THUMBWIDTH} gets replaced with the width of a thumbnail image
    * {TITLE} same as {POSTTITLE}
    * {URL} same as {FILELINK}
    * {WIDTH} gets replaced with width of the photo
    * {ICON} insert the icon for the attachment (for non-audio/image/video attachments only)


= Interoperability =
* If your mail client doesn't support setting the subject (nokia) you
  can do so by putting #your subject/title here# at the beginning of your message
* POP3,POP3-SSL,IMAP,IMAP-SSL now supported - last three require
  php-imap support
* The program understands enough about mime to not duplicate post
  if you send an HTML and plain text message
* Automatically confirms that you are installed correctly

== Frequently Asked Questions ==

= Postie won't connect to my mailserver. Why Not? =

Make sure the port you are using is open. For example, bluehost seems to block
ports 993 and 995 (for pop3-ssl and imap-ssl) by default. I have heard that
you can request that they open them for you ( you might have to pay extra). 

You can check for open ports with the following command on your server:
`netstat -ln|grep -E ':::(993|995|143)'`

If nothing shows up, then the ports are not open.

= How can I get postie to display inline images? =

Make sure that you send e-mail formatted as html (richtext), and set postie to
prefer html messages (in the message tab of the postie settings)

= Mail is not showing up right when I send html (rich formatted) e-mail! =

Make sure you set the preferred text type to html

= Do I need to any code to my theme for postie to work? =

No. 

= I read somewhere to add an iframe to my footer. Should I do this? =

No. Do not add an iframe in your footer to get postie to check mail
periodically. To check e-mail periodically, either set-up a cron job, or use
cronless postie. See installation instructions

= My mail host requires SSL, but postie will not allow me to select pop3-ssl or imap-ssl =

You must have php-imap installed on your server for this to work. Ask your
hosting provider about this.

= Can I use postie to check a gmail account? =

Yes. You can use either pop3-ssl or imap-ssl with a gmail account. Before
attempting to use with postie, make sure that you enable pop or imap in your
gmail preferences.

* Pop3 settings:
    * protocol - pop3-ssl
    * server - pop.gmail.com
    * port - 995
    * userid - your username (e.g. if your e-mail address is foo@gmail.com,
      this would be just foo)
    * password - your password 
* IMAP settings:
    * protocol - imap-ssl
    * server - imap.gmail.com
    * port - 993
    * userid - your username (e.g. if your e-mail address is foo@gmail.com,
      this would be just foo)
    * password - your password 

= Can I use postie with GoDaddy hosting? =

Yes, but you must use an email set up using the GoDaddy email service and the following settings:

* protocol - pop3
* server - pop.secureserver.net
* port - 110
* userid - xxxxx@yourdomain
* password - your password 

GoDaddy hosting does not allow you to connect to non-GoDaddy mail servers like Gmail.

= My posts show up as being posted by 'admin' instead of me. Why? =

If your admin account is linked to bar@gmail.com, and you send mail from
bar@gmail.com, it will show up as being posted by admin. If you have a
wordpress user named "John Doe", which is linked to johndoe@gmail.com, make
sure that you send e-mails from johndoe@gmail.com. It doesn't matter which
e-mail address postie is checking. That is, if you send mail from
johndoe@gmail.com to foo@gmail.com, it gets posted as "John Doe". 

If you send an e-mail to your postie address from an e-mail address that is no
t linked to a wordpress user, it will get posted as admin.

= Images aren't showing up at all? =

There are a couple possible reasons for this. First, check to see if you can
add an image through wordpress's normal posting mechanism. If not, then there
is probably 1 or 2 problems:
1. Your server does not have the php-gd library installed. Ask your hosting
provider about this.

2. Your wp-content/uploads directory is not writable by the webserver. Make
sure that it is

= Can I delete the wp-files directory needed by postie version <1.3.0? =

If you have posts published already by older versions of postie, getting rid
of those directories will delete any files you might have had associated with
those old posts. If you don't have any such posts, then you can safely delete
them.

= How can I get rid of stupid stuff my e-mail provider adds to my messages? =

To strip off stuff that they add at the beginning of a message, start your
post with :start

To strip off stuff that they add at the end of a message, end your
post with :end

= How can I add custom attachment icons? =

Simply upload the icons you want to the postie/icons/custom directory. You
must name the icons according to the following scheme:
`{filetype}-{size}.png`

NOTE THAT ANY CUSTOM FILES ADDED TO THE POSTIE DIRECTORY WILL BE DELETED IF YOU 
USE THE NORMAL WORDPRESS UPGRADE PROCESS.

For example, for word documents, you could use:

`doc-32.png`

for a 32x32 pixel icon. (You can actually create any size icon you want, but
if you name it 32, then it will only be used if you select to use size 32
icons)

See the other directories in icons for more examples.

Currently the following filetypes are supported:

* doc - microsoft word (including docx)
* ppt - microsoft powerpoint (including pptx)
* xls - microsoft excel (including xlsx)
* numbers - iWork numbres spreadsheet
* pages - iWork pages document
* key - iWork keynote presentation
* pdf - portable document format
* rtf - rich text format
* txt - plain text document

= Can I add special text to the body of the post when using postie? =

Yes. You can create your own function, and use the postie_post filter.
Two short examples are included in the filterPostie.php.sample file

= Can I add special text to the title of the post when using postie? =

Yes. You can create your own function, and use the postie_post filter.
Two short examples are included in the filterPostie.php.sample file

= Can I select tags or categories based on the content of the e-mail? =

Yes. You can create your own function, and use the postie_post filter.
See the filterPostie.php.sample file for examples.

= Is the IMAP extension required for postie? =

The IMAP extension is not required, but it is strongly recommended, especially
is you are using non-English text. There is more information on php.net about
installing the IMAP extension. If you have control over your server, it is
often not hard to install. 

On Ubuntu, try
`sudo apt-get install php5-imap`

On Fedora, try
`sudo yuminstall php-imap`

The IMAP extension is known to be installed on the following popular webhosts:
* Dreamhost

= How can I embed youtube or vimeo videos? =

Simply put the url in the body of your e-mail. (Make sure that you have the
option to convert url into links turned on)

= Something is going wrong, how do I see what is happening? = 

Add the following lines to your wp-config.php file
`define('WP_DEBUG', true);
define('WP_DEBUG_LOG', true);
define('POSTIE_DEBUG', true);`

Various errors, warning and informational will be written to the wp-content\debug.log
file. There may also be all sorts of warnings and messages in your site as well
depending on how well behaved your other plugins and themes are, so you will not
want to leave these settings set to true all the time.

= Why doesn't Postie automatically publish my emails when running manually works =

WordPress cron (which Postie relies on) doesn't run unless a page is accessed on the 
site. So if you send an email, but nobody accesses the site for 3 days Postie won't 
be given the chance to fetch the email and publish the post.

To ensure that Postie runs smoothly on a low or no volume site you need to ensure that 
a page gets hit (any page is fine). Use something like cron + curl on Linux or install 
curl on Windows and use the Scheduled Tasks control panel. If you are using a hosting company
that doesn't allow you access to cron you can use a service like 
<a href="https://www.setcronjob.com/" target="_blank">SetCronJob</a>.

It is also possible to turn the WordPress cron off. Please make sure something like
`define('DISABLE_WP_CRON', true);` is not in your wp-config.php file.
== Upgrade Notice ==

= 1.5.3 =
* Postie can now set the first image in an email to be the "Featured" image. There is a new setting "Use First Image as Featured Image" which is off by default.
* Postie now supports Use Transport Layer Security (TLS)

= 1.5.0 =
* New filter postie_filter_email. Used to map "from" to any other email. Allows custom user mapping.

= 1.4.41 =
Post format is now supported. You can specify any of the WordPress supported post formats using the Post type syntax.
Post status can now be specified using the status: tag.
Post status setting was renamed to Default Post Status and moved to the Message tab.

= 1.4.18 =
Many method names have been changed. Any custom filters may need to be updated.

= 1.4.10 =
All script, style and body tags are stripped from html emails.

= 1.4.6 =
Attachments are now processed in the order they were attached.

== CHANGELOG ==
= 1.5.7 (2013.05.09) =
* Fixed bug where the admin user was not getting set as author in some cases.
* Fixed bug where file names were not being sanitized.
* Added setting to disable thumbnail generation.
* Updated the default signature patterns and help text.

= 1.5.6 (2013-05-07) =
* Fixed bug where default post format was empty when the theme didn't support any formats.
* Removed all hard coded references to wp-content.
* Fixed bug where caption wasn't being filled out in templates.
* Updated direct DB access to use standard methods.
* Added attachment template for non-audio/image/video attachments.
* Added additional checks to the test screen.
* Added new template variable {ICON}

= 1.5.5 (2013.05.02) =
* Added ability to run a manual check with debugging output.
* Added support for default post type.
* Added version number to settings page.
* Changed Admin User to Default Poster, clarified help text and changed input to combo box of valid values.
* Fixed bug where commas (,) were not allowed in signature detection.
* Fixed some CSS issues on admin page.

= 1.5.4 (2013.05.01) =

docs/Changes.txt
== Upgrade Notice ==

= 1.5.3 =
* Postie can now set the first image in an email to be the "Featured" image. There is a new setting "Use First Image as Featured Image" which is off by default.
* Postie now supports Use Transport Layer Security (TLS)

= 1.5.0 =
* New filter postie_filter_email. Used to map "from" to any other email. Allows custom user mapping.

= 1.4.41 =
Post format is now supported. You can specify any of the WordPress supported post formats using the Post type syntax.
Post status can now be specified using the status: tag.
Post status setting was renamed to Default Post Status and moved to the Message tab.

= 1.4.18 =
Many method names have been changed. Any custom filters may need to be updated.

= 1.4.10 =
All script, style and body tags are stripped from html emails.

= 1.4.6 =
Attachments are now processed in the order they were attached.

== CHANGELOG ==
= 1.5.7 (2013.05.09) =
* Fixed bug where the admin user was not getting set as author in some cases.
* Fixed bug where file names were not being sanitized.
* Added setting to disable thumbnail generation.
* Updated the default signature patterns and help text.

= 1.5.6 (2013-05-07) =
* Fixed bug where default post format was empty when the theme didn't support any formats.
* Removed all hard coded references to wp-content.
* Fixed bug where caption wasn't being filled out in templates.
* Updated direct DB access to use standard methods.
* Added attachment template for non-audio/image/video attachments.
* Added additional checks to the test screen.
* Added new template variable {ICON}

= 1.5.5 (2013.05.02) =
* Added ability to run a manual check with debugging output.
* Added support for default post type.
* Added version number to settings page.
* Changed Admin User to Default Poster, clarified help text and changed input to combo box of valid values.
* Fixed bug where commas (,) were not allowed in signature detection.
* Fixed some CSS issues on admin page.

= 1.5.4 (2013.05.01) =
* Added support for default post format
* Fixed bug where replies were getting attached to the wrong post
* Fixed bug where tags were not being detected correctly in html emails
* Fixed bug in reply detection

= 1.5.3 (2013.04.13) =
* Added support for Featured Images
* Added support for Use Transport Layer Security (TLS)
* Updated postie_filter_email filter to get unprocessed email address

= 1.5.2 (2013.04.12) =
* Fixed bug in post type/format detection when no valid post type/format was found
* Workaround for WP bug when POP3 account has no waiting messages
* Fixed bug where cron was running postie_check on every page load

= 1.5.1 (2013.04.10) =
* Turned on POP3 debug logging if POSTIE_DEBUG is TRUE.
* Disable autocomplete on some setup fields
* Fixed bug where confirmation emails were not being sent to authors
* Fixed bug where post were not saved if the default admin user didn't exist and the from user was not a WordPress user

= 1.5.0 (2013.04.05) =
* Apply Postie Time Correction to date: command
* Add support for Post Formats
* Add support for Post Status
* Add warning if Admin username is invalid
* Fixed bug where date: was not always being detected in html emails
* Improved handling of attachments and mapping to file extensions for template selection
* Video templates now include scale="tofit"
* Add support for older png mime type
* New filter postie_filter_email. Used to map "from" to any other email. Allows custom user mapping.


= 1.4.40 (2013.03.18) =
* Fixed bug where categories specified by ID were not being correctly identified

= 1.4.39 (2013.03.14) =
* Fixed bug where Postie supplied schedules were not always being correctly added to the cron

= 1.4.38 (2013.03.12) =
* Improved POP3 configuration test
* Fixed bug where :start and :end were removing commands like tags: and date: before they got a chance to be processed

= 1.4.37 (2013.03.11) =
* Fixed bug in tag handling
* Fixed bug in category detection
* Worked around a bug in WordPress that was mangling filenames that are not in ASCII, i.e. Arabic, Hebrew, Cyrillic, etc.

= 1.4.36 (2013.03.07) =
* Removed some debugging code from filters and hooks
* Fixed bug where the date command was being ignored
* Added 'quicktime' as a default type for the video1 template

= 1.4.35 (2013.02.22) =
* Consolidated logic for load configuration settings. Fixes bug where new settings were not having their defaults set properly.
* Fixed bug where attachment file name was not being correctly detected.

= 1.4.34 (2013.02.07) =
* Fixed bug in new category logic

= 1.4.33 (2013.02.05) =
* Fixed bug where non-category taxonomy was being selected as the post category.
* Added option to force categories to match exactly.
* Added logic to skip text attachments named "ATT00001.txt" where the numbers can be any sequence. This
  improves the previous fix by detecting all attachments added, not just the first one.

= 1.4.32 (2013.01.29) =
* Fixed bug in detecting need for imap extension.
* Added additional selections for "Maximum number of emails to process"
* Added logic to skip text attachments named "ATT00001.txt" which are added by MS Exchange virus scanning.
* Added option to check for email every 5 minutes.

= 1.4.31 (2013.01.25) =
* Enhanced category detection to be compatible with Polylang plugin.
* Enhanced prerequisite detection.
* Now using wp_mail() instead of mail()
* WordPress 3.5.1 compatibility

= 1.4.30 (2013.01.22) =
* Fixed bug that caused activation to fail or show a blank page.
* Fixed bug where WP media upload rejects a file.
* Added a check for mbstring.
* Fixed bug when attachment names were only supplied via d_parameters.

= 1.4.29 (2013.01.19) =
* Fixed bug where manually running Postie worked, but calling get_mail.php directly did not.

= 1.4.28 (2013.01.18) =

postie.php
<?php

/*
  Plugin Name: Postie
  Plugin URI: http://PostiePlugin.com/
  Description: Signifigantly upgrades the posting by mail features of Word Press (See <a href='options-general.php?page=postie/postie.php'>Settings and options</a>) to configure your e-mail settings. See the <a href='http://wordpress.org/extend/plugins/postie/other_notes'>Readme</a> for usage. Visit the <a href='http://wordpress.org/support/plugin/postie'>postie forum</a> for support.
  Version: 1.5.7
  Author: Wayne Allen
  Author URI: http://allens-home.com/
  License: GPL2
 */

/*  Copyright (c) 2012  Wayne Allen  (email : wayne@allens-home.com)

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License, version 2, as
  published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

/*
  $Id: postie.php 710554 2013-05-09 19:28:59Z WayneAllen $
 */

define('POSTIE_VERSION', '1.5.7');
define("POSTIE_ROOT", dirname(__FILE__));
define("POSTIE_URL", WP_PLUGIN_URL . '/' . basename(dirname(__FILE__)));

//register the hooks early in the page in case some method needs the result of one of them (i.e. cron_schedules)
add_action('init', 'disable_kses_content', 20);
add_action('check_postie_hook', 'check_postie');

add_filter('whitelist_options', 'postie_whitelist');
add_filter('cron_schedules', 'postie_more_reccurences');

register_activation_hook(__FILE__, 'activate_postie');
register_activation_hook(__FILE__, 'postie_cron');
register_deactivation_hook(__FILE__, 'postie_decron');

function postie_loadjs_add_page() {
    $postiepage = add_options_page('Postie', 'Postie', 'manage_options', POSTIE_ROOT . '/postie.php', 'postie_loadjs_options_page');
    add_action("admin_print_scripts-$postiepage", 'postie_loadjs_admin_head');
}

function postie_loadjs_options_page() {
    require_once POSTIE_ROOT . '/config_form.php';
}

function postie_loadjs_admin_head() {
    wp_enqueue_script('loadjs', plugins_url('js/simpleTabs.jquery.js', __FILE__));
    echo '<link type="text/css" rel="stylesheet" href="' . plugins_url('css/style.css', __FILE__) . "\"/>\n";
    echo '<link type="text/css" rel="stylesheet" href="' . plugins_url('css/simpleTabs.css', __FILE__) . "\"/>\n";
}

if (isset($_GET["postie_read_me"])) {
    include_once(ABSPATH . "wp-admin/admin.php");
    $title = __("Edit Plugins");
    $parent_file = 'plugins.php';
    include(ABSPATH . 'wp-admin/admin-header.php');
    postie_ShowReadMe();
    include(ABSPATH . 'wp-admin/admin-footer.php');
}
//Add Menu Configuration
if (is_admin()) {
    require_once(dirname(__FILE__) . DIRECTORY_SEPARATOR . "postie-functions.php");
    add_action('admin_init', 'postie_admin_settings');
    add_action('admin_menu', 'postie_loadjs_add_page');
    if (function_exists('load_plugin_textdomain')) {

        function postie_load_domain() {
            $plugin_dir = WP_PLUGIN_DIR . '/' . basename(dirname(__FILE__));
            load_plugin_textdomain('postie', $plugin_dir . "/languages/", basename(dirname(__FILE__)) . '/languages/');
        }

        add_action('init', 'postie_load_domain');
    }
    postie_warnings();
}

/*
 * called by WP when activating the plugin
 * Note that you can't do any output during this funtion or activation
 * will fail on some systems. This means no DebugEcho, EchoInfo or DebugDump.
 */

function activate_postie() {
    static $init = false;
    $options = config_Read();

    if ($init) {
        return;
    }

    if (!$options) {
        $options = array();
    }
    $default_options = config_GetDefaults();
    $old_config = array();

    $result = config_GetOld();
    if (is_array($result)) {
        foreach ($result as $key => $val) {
            $old_config[strtolower($key)] = $val;
        }
    }

    // overlay the options on top of each other:
    // the current value of $options takes priority over the $old_config, which takes priority over the $default_options
    $options = array_merge($default_options, $old_config, $options);
    $options = config_ValidateSettings($options);
    update_option('postie-settings', $options);
    $init = true;
}

/**
 * set up actions to show relevant warnings, 
 * if mail server is not set, or if IMAP extension is not available
 */
function postie_warnings() {

    $config = config_Read();

    if ((empty($config['mail_server']) ||
            empty($config['mail_server_port']) ||
            empty($config['mail_userid']) ||
            empty($config['mail_password'])
            ) && !isset($_POST['submit'])) {

        function postie_enter_info() {
            echo "<div id='postie-info-warning' class='updated fade'><p><strong>" . __('Postie is almost ready.', 'postie') . "</strong> "
            . sprintf(__('You must <a href="%1$s">enter your email settings</a> for it to work.', 'postie'), "options-general.php?page=postie/postie.php")
            . "</p></div> ";
        }

        add_action('admin_notices', 'postie_enter_info');
    }

    $p = strtolower($config['input_protocol']);
    if (!function_exists('imap_mime_header_decode') && ($p == 'imap' || $p == 'imap-ssl' || $p == 'pop-ssl')) {

        function postie_imap_warning() {
            echo "<div id='postie-imap-warning' class='error'><p><strong>";
            echo __('Warning: the IMAP php extension is not installed. Postie can not use IMAP, IMAP-SSL or POP-SSL without this extension.', 'postie');
            echo "</strong></p></div>";
        }

        add_action('admin_notices', 'postie_imap_warning');
    }
    if ($p == 'pop3' && $config['email_tls']) {

        function postie_tls_warning() {
            echo "<div id='postie-lst-warning' class='error'><p><strong>";
            echo __('Warning: The POP3 connector does not support TLS.', 'postie');
            echo "</strong></p></div>";
        }

        add_action('admin_notices', 'postie_tls_warning');
    }

    if (!function_exists('mb_detect_encoding')) {

        function postie_mbstring_warning() {
            echo "<div id='postie-mbstring-warning' class='error'><p><strong>";
            echo __('Warning: the Multibyte String php extension (mbstring) is not installed. Postie will not function without this extension.', 'postie');
            echo "</strong></p></div>";
        }

        add_action('admin_notices', 'postie_mbstring_warning');
    }

    if (!function_exists('get_user_by')) {
        include ABSPATH . 'wp-includes/pluggable.php';
    }
    $adminuser = get_user_by('login', $config['admin_username']);
    if ($adminuser === false) {

        function postie_adminuser_warning() {
            echo "<div id='postie-mbstring-warning' class='error'><p><strong>";
            echo __('Warning: the Admin username is not a valid WordPress login. Postie may reject emails if this is not corrected.', 'postie');
            echo "</strong></p></div>";
        }

        add_action('admin_notices', 'postie_adminuser_warning');
    }
}

function disable_kses_content() {
    remove_filter('content_save_pre', 'wp_filter_post_kses');
}

function postie_whitelist($options) {
    $added = array('postie-settings' => array('postie-settings'));
    $options = add_option_whitelist($added, $options);
    return $options;
}

//don't use DebugEcho or EchoInfo here as it is not defined when called as an action
function check_postie() {
    error_log("check_postie");

    $fullurl = plugins_url("get_mail.php", __FILE__);
    preg_match("/https?:\/\/(.[^\/]*)(.*)/i", $fullurl, $matches);
    $host = $matches[1];

    $url = "";
    if (isset($matches[2])) {
        $url = $matches[2];
    }

    $port = is_ssl() ? 443 : 80;

    $fp = fsockopen($host, $port, $errno, $errstr);
    if ($fp) {
        fputs($fp, "GET $url HTTP/1.0\r\n");
        fputs($fp, "User-Agent:  Cronless-Postie\r\n");
        fputs($fp, "Host: $host\r\n");
        fputs($fp, "\r\n");
        $page = '';
        while (!feof($fp)) {
            $page.=fgets($fp, 128);
        }
        fclose($fp);
    } else {
        error_log("Cannot connect to server on port $port. Please check to make sure that this port is open on your webhost. Additional information: $errno: $errstr");
    }
}

function postie_cron($interval = false) {
    //Do not echo output in filters, it seems to break some installs
    //error_log("postie_cron: setting up cron task: $interval");
    //$schedules = wp_get_schedules();
    //error_log("postie_cron\n" . print_r($schedules, true));

    if (!$interval) {
        $config = config_Read();
        $interval = $config['interval'];
        //error_log("postie_cron: setting up cron task from config: $interval");
    }
    if (!$interval || $interval == '') {
        $interval = 'hourly';
        //error_log("Postie: setting up cron task: defaulting to hourly");
    }
    if ($interval == 'manual') {
        postie_decron();
        //error_log("postie_cron: clearing cron (manual)");
    } else {
        if ($interval != wp_get_schedule('check_postie_hook')) {
            postie_decron(); //remove existing
            //try to create the new schedule with the first run in 5 minutes
            if (false === wp_schedule_event(time() + 5 * 60, $interval, 'check_postie_hook')) {
                //error_log("postie_cron: Failed to set up cron task: $interval");
            } else {
                //error_log("postie_cron: Set up cron task: $interval");
            }
        } else {
            //error_log("postie_cron: OK: $interval");
            //don't need to do anything, cron already scheduled
        }
    }
}

function postie_decron() {
    //error_log("postie_decron: clearing cron");
    wp_clear_scheduled_hook('check_postie_hook');
}

/* here we add some more cron options for how often to check for e-mail */

function postie_more_reccurences($schedules) {
    //Do not echo output in filters, it seems to break some installs
    //error_log("postie_more_reccurences: setting cron schedules");
    $schedules['weekly'] = array('interval' => (60 * 60 * 24 * 7), 'display' => __('Once Weekly'));
    $schedules['twiceperhour'] = array('interval' => 60 * 30, 'display' => __('Twice per hour'));
    $schedules['tenminutes'] = array('interval' => 60 * 10, 'display' => __('Every 10 minutes'));
    $schedules['fiveminutes'] = array('interval' => 60 * 5, 'display' => __('Every 5 minutes'));

    return $schedules;
}

?>